variable-precision float operations can differ depending on optimization levels

RalfJung opened the pull request:
Follow-up to https://github.com/rust-lang/rust/pull/121793 and https://github.com/rust-lang/rust/pull/118217 that accounts for optimizations changing the precision of these functions.

Fixes https://github.com/rust-lang/rust/issues/109118
Fixes https://github.com/rust-lang/rust/issues/71355

RalfJung committed:
variable-precision float operations behave non-deterministically

Muon commented:
I don't think this is a good idea. Why should libm functions be privileged in this way? They ought to be treated as ordinary library calls.

RalfJung commented:
This describes the current reality. Feel free to make a proposal for changing our semantics, but I don't think that should block documenting the current semantics. In particular this is future compatible with changing them later to be optimization-independent.

Also note that we are not really privileging anything here. What we are effectively saying is that these functions are returning a result with arbitrary nondeterministic precision. It then follows as a consequence that the precision can depend on host, target, compiler version, flags, or literally anything else.

Maybe we should mention the non-determinism more explicitly.

RalfJung commented:
Yeah, something like that -- I updated the PR.

The section title "Platform-specific precision" doesn't make a lot of sense any more now, since precision is not "specific" to any platform, so I changed that as well.

cuviper commented:
Works for me!

@bors r+ rollup

bors commented:
:pushpin: Commit ff2ff9790436ca16a69c2237a5a8b8a88c69cdbf has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper ff2ff9790436ca16a69c2237a5a8b8a88c69cdbf -->
<!-- homu: {"type":"Approved","sha":"ff2ff9790436ca16a69c2237a5a8b8a88c69cdbf","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Muon commented:
Ah, I like this a lot better. At least the danger of using them is much clearer now.

RalfJung commented:
@bors r-
